Hammer Updates (#886)
ucb-bar/chipyard: docs/VLSI/Tutorial.rst

* Genus, Innovus, and Calibre licenses
* For ASAP7 specifically:

>   * Download the `ASAP7 PDK v1p5 <http://asap.asu.edu/asap/>`__ tarball to a directory of choice but do not extract it. The tech plugin is configured to extract the PDK into a cache directory for you. Note: v1p5 of the PDK is not publicly available, and you will need to contact the developers for it. The v1p7 version currently has too many bugs.

Reviewer: Perhaps clarify to say the `v1p7 PDK version is currently too buggy to be integrated into our flow`?

>   * Download the `ASAP7 PDK v1p5 <http://asap.asu.edu/asap/>`__ tarball to a directory of choice but do not extract it. The tech plugin is configured to extract the PDK into a cache directory for you. Note: v1p5 of the PDK is not publicly available, and you will need to contact the developers for it. The v1p7 version that is `publicly released <https://github.com/The-OpenROAD-Project/asap7>`__ currently has too many bugs to be integrated into our flow.

Reviewer 2: Could this be written a bit more diplomatically? (the "has too many bugs to be integrated into our flow" part).
For example, "has some critical issues involving X, Y, Z which currently prevent it from be integrated into the Hammer flow"

Author: I don't want to enumerate what's already in the issue list...

>   * Download the `ASAP7 PDK v1p5 <http://asap.asu.edu/asap/>`__ tarball to a directory of choice but do not extract it. The tech plugin is configured to extract the PDK into a cache directory for you. Note: v1p5 of the PDK is not publicly available, and you will need to contact the developers for it. The v1p7 version that is `publicly released <https://github.com/The-OpenROAD-Project/asap7>`__ currently has too many bugs to be integrated into our flow.

Reviewer 2: ```suggestion
  * Download the `ASAP7 PDK v1p5 <http://asap.asu.edu/asap/>`__ tarball to a directory of choice but do not extract it. The tech plugin is configured to extract the PDK into a cache directory for you. Note: v1p5 of the PDK is not publicly available, and you will need to contact the developers for it. The v1p7 version that is `publicly released <https://github.com/The-OpenROAD-Project/asap7>`__ currently has several critical issues which prevent it from being fully integrated into the Hammer flow.
```

Building the Design
--------------------

> To elaborate the ``TinyRocketConfig`` and set up all prerequisites for the build system to push just the accelerator + hard macro through the flow:

Reviewer 2: I'm confused - if we switched from Sha3RocketConfig to TinyRocketConfig, is there still an accelerator and hard macro?

> To elaborate the ``TinyRocketConfig`` and set up all prerequisites for the build system to push just the accelerator + hard macro through the flow:

Reviewer 2: ```suggestion
To elaborate the ``TinyRocketConfig`` and set up all prerequisites for the build system to push the design and SRAM macros through the flow:
```

To elaborate the ``TinyRocketConfig`` and set up all prerequisites for the build system to push just the accelerator + hard macro through the flow:

.. code-block:: shell

    make buildfile MACROCOMPILER_MODE='--mode synflops' CONFIG=Sha3RocketConfig VLSI_TOP=Sha3AccelwBB

> The ``MACROCOMPILER_MODE='--mode synflops'`` is needed because the ASAP7 process does not yet have a memory compiler, so flip-flop arrays are used instead. This will dramatically increase the synthesis runtime if your design has a lot of memory state (e.g. large caches). This change is automatically inferred by the makefile but is included here for completeness.

Reviewer 2: So ASAP7 SRAMs work of-out-the-box now?

Author: They're still dummy blocks, but they are targetable for behavioral sim, synthesis, and P&R only.

.. code-block:: shell

    make drc CONFIG=TinyRocketConfig
    ./build/drc-rundir/generated-scripts/view-drc
    make lvs CONFIG=TinyRocketConfig
    ./build/lvs-rundir/generated-scripts/view-lvs

Some DRC errors are expected from this PDK, as explained in the `ASAP7 plugin readme <https://github.com/ucb-bar/hammer/tree/master/src/hammer-vlsi/technology/asap7>`__.

> Furthermore, the dummy SRAMs that are provided do not have any geometry inside, so will certainly cause DRC and LVS errors.

Reviewer 2: Is there any discussion of these "dummy SRAMs" anywhere?

Simulation
^^^^^^^^^^
Simulation with VCS is supported, and can be run at the RTL- or gate-level (post-synthesis and P&R). The simulation infrastructure as included here is intended for running RISC-V binaries on a Chipyard config. For example, for an RTL-level simulation:

.. code-block:: shell

    make sim-rtl CONFIG=TinyRocketConfig BINARY=$RISCV/riscv64-unknown-elf/share/riscv-tests/isa/rv64ui-p-simple

Post-synthesis and post-P&R simulations use the ``sim-syn`` and ``sim-par`` targets, respectively.

There are also ``-debug`` and ``-debug-timing``, which will instruct VCS to write a SAIF + VPD and do timing-annotated simulations, respectively. See the ``sim.mk`` file for all available targets.

> Note that for the ASAP7 example, gate-level simulations will timeout (the cause is being investigated).

Reviewer 2: timeout in which tool?

Author: VCS. I will do my best to investigate by tutorial time.

> Note that for the ASAP7 example, gate-level simulations will timeout (the cause is being investigated).

Reviewer 2: ```suggestion
Note that for the ASAP7 example, gate-level simulations will currently timeout.
```
